pkg/inttest/inttest.go

```diff
@@ -0,0 +1,298 @@
+// finished and return a client ready to interact with the container.
+package inttest
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	amqp "github.com/rabbitmq/amqp091-go"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/dhis2-sre/im-manager/internal/handler"
+	"github.com/dhis2-sre/im-manager/internal/server"
+	"github.com/dhis2-sre/im-manager/pkg/storage"
+	"github.com/gin-gonic/gin"
+	_ "github.com/lib/pq" // postgres driver
+	"github.com/orlangure/gnomock"
+	"github.com/orlangure/gnomock/preset/localstack"
+
+	"github.com/dhis2-sre/im-manager/pkg/config"
+	"github.com/orlangure/gnomock/preset/postgres"
+	"github.com/orlangure/gnomock/preset/rabbitmq"
+
+	"github.com/stretchr/testify/require"
+	"gorm.io/gorm"
+)
+
+// SetupDB creates a PostgreSQL container. Gorm is connected to the DB and runs the migrations.
+func SetupDB(t *testing.T) *gorm.DB {
+	t.Helper()
+
+	container, err := gnomock.Start(
+		postgres.Preset(
+			postgres.WithUser("im", "im"),
+			postgres.WithDatabase("test_im"),
+		),
+	)
+	require.NoError(t, err, "failed to start DB")
+	t.Cleanup(func() { require.NoError(t, gnomock.Stop(container), "failed to stop DB") })
+
+	db, err := storage.NewDatabase(config.Postgresql{
+		Host:         container.Host,
+		Port:         container.DefaultPort(),
+		Username:     "im",
+		Password:     "im",
+		DatabaseName: "test_im",
+	})
+	require.NoError(t, err, "failed to setup DB")
+	return db
+}
+
+// SetupS3 creates an S3 container (using localstack) with all the buckets and files in given path.
+func SetupS3(t *testing.T, path string) *S3Client {
+	t.Helper()
+
+	container, err := gnomock.Start(
+		localstack.Preset(
+			localstack.WithServices(localstack.S3),
+			localstack.WithS3Files(path),
+			localstack.WithVersion("2.1.0"),
+		),
+	)
+	require.NoError(t, err, "failed to start S3")
+	t.Cleanup(func() { require.NoError(t, gnomock.Stop(container), "failed to stop S3") })
+
+	return &S3Client{
+		Client: s3.NewFromConfig(
+			aws.Config{
+				Region: "eu-west-1",
+				EndpointResolverWithOptions: aws.EndpointResolverWithOptionsFunc(func(service, region string, options ...interface{}) (aws.Endpoint, error) {
+					return aws.Endpoint{
+						URL:           fmt.Sprintf("http://%s/", container.Address(localstack.APIPort)),
+						SigningRegion: "eu-west-1",
+					}, nil
+				}),
+			},
+			func(o *s3.Options) {
+				o.UsePathStyle = true
+			},
+		),
+	}
+}
+
+// S3Client allows making requests to S3. It does so by wrapping an S3.Client. Access the actual
+// S3.Client for specific use cases where our defaults don't work.
+type S3Client struct {
+	Client *s3.Client
+}
+
+func (sc *S3Client) GetObject(t *testing.T, bucket, key string) []byte {
+	t.Helper()
+
+	oo, err := sc.Client.GetObject(context.TODO(), &s3.GetObjectInput{
```

Comment:
`oo` looks a bit confusing and cryptic, can we use something more descriptive name?

Reply:
its ok as you can immediately forget about it in the next line 😋 but I'll change it if you give me a suggestion?

Reply:
called it `object` 😄